source/common/common/key_value_store_base.cc

```diff
@@ -59,43 +53,49 @@ bool KeyValueStoreBase::parseContents(absl::string_view contents,
       ENVOY_LOG(warn, error);
       return false;
     }
-    store.emplace(std::string(key.value()), std::string(value.value()));
+    addOrUpdate(key.value(), value.value());
   }
   return true;
 }
 
-void KeyValueStoreBase::addOrUpdate(absl::string_view key, absl::string_view value) {
-  store_.insert_or_assign(key, std::string(value));
+void KeyValueStoreBase::addOrUpdate(absl::string_view key_view, absl::string_view value_view) {
+  ENVOY_BUG(!under_iterate_, "addOrUpdate under the stack of iterate");
+  std::string key(key_view);
+  std::string value(value_view);
+  // Attempt to insert the entry into the store. If it already exists, remove
+  // the old entry and insert the new one so it will be in the proper place in
+  // the linked list.
+  if (!store_.emplace(key, value).second) {
+    store_.erase(key);
+    store_.emplace(key, value);
+  }
+  if (max_entries_ && store_.size() > max_entries_) {
+    store_.pop_front();
+  }
   if (!flush_timer_->enabled()) {
     flush();
   }
 }
 
 void KeyValueStoreBase::remove(absl::string_view key) {
-  store_.erase(key);
+  ENVOY_BUG(!under_iterate_, "remove under the stack of iterate");
+  store_.erase(std::string(key));
   if (!flush_timer_->enabled()) {
     flush();
   }
 }
 
 absl::optional<absl::string_view> KeyValueStoreBase::get(absl::string_view key) {
-  auto it = store_.find(key);
+  auto it = store_.find(std::string(key));
   if (it == store_.end()) {
     return {};
   }
   return it->second;
 }
 
 void KeyValueStoreBase::iterate(ConstIterateCb cb) const {
-#ifndef NDEBUG
-  // When running in debug mode, verify we don't modify the underlying store
-  // while iterating.
-  absl::flat_hash_map<std::string, std::string> store_before_iteration = store_;
-  absl::Cleanup verify_store_is_not_modified = [this, &store_before_iteration] {
-    ASSERT(store_ == store_before_iteration,
-           "Expected iterate to not modify the underlying store.");
-  };
-#endif
+  under_iterate_ = true;
+  absl::Cleanup verify_store_is_not_modified = [this] { under_iterate_ = false; };
```

Comment:
This cleanup function doesn't seem to verify anything? Maybe `reset_iteration_state` or something like that?